sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +35,112 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	sdev = spcm->sdev;
+	info = spcm->time_info;
+
+	if (!host_copier || !dai_copier) {
+		dev_err(sdev->dev, "failed to find host copier %p or dai copier%p\n",
+			host_copier, dai_copier);
+		return;
+	}
+
+	if (spcm->substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		u64 stream_start_position;
+		u32 dai_sample_size;
+		u32 ch;
+
+		node_id = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
+		if (node_id >= SOF_IPC4_MAX_PIPELINE_REG_SLOTS) {
+			dev_err(sdev->dev, "invalid node_id 0x#%x\n", node_id);
+			return;
+		}
+
+		offset = offsetof(struct sof_ipc4_fw_registers, pipeline_regs) +
+				  node_id * sizeof(ppl_reg);
+		sof_mailbox_read(sdev, sdev->fw_info_box.offset + offset, &ppl_reg,
+				 sizeof(ppl_reg));
+		if (ppl_reg.stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+			dev_err(sdev->dev, "failed to get stream start offset\n");
+			return;
+		}
+
+		dev_dbg(sdev->dev, "stream_start_offset position %lld\n",
+			ppl_reg.stream_start_offset);
+
+		stream_start_position = ppl_reg.stream_start_offset;
+		ch = dai_copier->data.out_format.fmt_cfg;
+		ch = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(ch);
+		dai_sample_size = (dai_copier->data.out_format.bit_depth >> 3) * ch;
+		/* convert offset to sample count */
+		do_div(stream_start_position, dai_sample_size);
+		info->stream_start_offset = stream_start_position;
+	} else {
+		/* capture doesn't support stream_start_offset in fw */
+		info->stream_start_offset = 0;
+	}
+
+	info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_gpdma_reading_slots) +
+				    sdev->fw_info_box.offset;
+
+	for (i = 0; i < SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS; i++) {
+		info->llp_offset += i * sizeof(struct sof_ipc4_llp_reading_slot);
+		sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+
+		if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+			break;
+	}
+
+	if (i >= SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS) {
+		/* if it is not found, check aggregated sdw slot */
+		info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_sndw_reading_slots) +
+						sdev->fw_info_box.offset;
+		for (i = 0; i < SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS - 1; i++) {
+			info->llp_offset += i * sizeof(struct sof_ipc4_llp_reading_slot);
+			sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+
+			if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+				break;
+		}
+	}
+
+	/* no llp support */
+	if (i >= SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS - 1) {
+		dev_warn(sdev->dev, "no llp found, fall back to default path");
+		info->llp_offset = 0;
+	}
+}
```

Comment:
Do we need this patch @RanderWang ?

It doesn't seem to be used for the delay calculation?

Reply:
@plbossart  two points   
 (1) get dai position from memory window based on llp offset. We can use dai position from pcm_pointer so it can be removed
(2) get stream_start_position for multiple-stream mixing. if we don't use it we can't deal with mixer case.

Reply:
Add more: for (1) we can only get dai position for HDA since the position can be gotten from host HDA registers, but it can't work for I2S or SDW, so we still have to get dai position from memory window

Reply:
@RanderWang the question was "why do we need to read such registers in a workqueue function" ?
That seems counter-intuitive for values that are time-sensitive.

Reply:
yes, I guess you have such concern.  We can only get invalid stream_start_offset  when pipeline is  just triggered start since FW is scheduling pipelines (two or more pipelines) and stream_start_offset is not ready so we need to wait for a while. And ASoC will trigger sof , sdw controller and codec to start, so audio dma doesn't transfer data until controller and codec are ready.

I am working on another method without workqueue that in delay function we will check stream_start_offset . If it is valid we will use it or read it again from memory window. In my experiment it is only invalid in the first delay call